src/hdmf/common/table.py

```diff
@@ -470,37 +487,39 @@ def __eq__(self, other):
     def add_column(self, **kwargs):
-        name = getargs('name', kwargs)
-        for col in self.__columns__:
-            if col['name'] == name:  # column has not been added but is pre-specified
-                msg = "column '%s' already exists in %s '%s'" % (name, self.__class__.__name__, self.name)
-                raise ValueError(msg)
-
-        self._add_column(**kwargs)
-
-    @docval({'name': 'name', 'type': str, 'doc': 'the name of this VectorData'},
-            {'name': 'description', 'type': str, 'doc': 'a description for this column'},
-            {'name': 'data', 'type': ('array_data', 'data'),
-             'doc': 'a dataset where the first dimension is a concatenation of multiple vectors', 'default': list()},
-            {'name': 'table', 'type': (bool, 'DynamicTable'),
-             'doc': 'whether or not this is a table region or the table the region applies to', 'default': False},
-            {'name': 'index', 'type': (bool, VectorIndex, 'array_data'),
-             'doc': 'whether or not this column should be indexed', 'default': False})
-    def _add_column(self, **kwargs):
         """
         Add a column to this table.
 
         If data is provided, it must contain the same number of rows as the current state of the table.
 
-        :raises ValueError
+        :raises ValueError: if the column has already been added to the table
         """
-        name, data = getargs('name', 'data', kwargs)
+        name, data, description = getargs('name', 'data', 'description', kwargs)
         index, table = popargs('index', 'table', kwargs)
 
         if name in self.__colids:  # column has already been added
             msg = "column '%s' already exists in %s '%s'" % (name, self.__class__.__name__, self.name)
             raise ValueError(msg)
 
+        if name in self.__uninit_cols:  # column is a predefined optional column from the spec
+            # check the given values against the predefined optional column spec. if they do not match, raise a warning
+            # and ignore the given arguments. users should not be able to override these values
+            spec_table = self.__uninit_cols[name].get('table', False)
+            if table != spec_table:
+                msg = ("Column '%s' is predefined in %s with table=%s which does not match the entered "
+                       "table argument. The entered table argument will be ignored."
+                       % (name, self.__class__.__name__, spec_table))
+                warn(msg)
+                table = spec_table
+
+            spec_index = self.__uninit_cols[name].get('index', False)
+            if index != spec_index:
+                msg = ("Column '%s' is predefined in %s with index=%s which does not match the entered "
+                       "index argument. The entered index argument will be ignored."
+                       % (name, self.__class__.__name__, spec_index))
+                warn(msg)
+                index = spec_index
+
```

Comment:
Note that this behavior differs from when columns are specified via `DynamicTable.__init__`. In that case, the given columns override the predefined columns. I don't think that overriding should be allowed, or at least a warning should be raised. See new line #327 for the TODO.